rbs/MethodTypeToParserNode.cc

```diff
@@ -52,13 +52,109 @@ core::LocOffsets adjustNameLoc(const RBSDeclaration &declaration, rbs_node_t *no
     return declaration.typeLocFromRange(range);
 }
 
-unique_ptr<parser::Node> handleAnnotations(unique_ptr<parser::Node> sigBuilder, const vector<Comment> &annotations) {
+bool isSelfOrKernel(parser::Node *node) {
+    if (parser::isa_node<parser::Self>(node)) {
+        return true;
+    }
+
+    if (auto constant = parser::cast_node<parser::Const>(node)) {
+        return constant->name == core::Names::Constants::Kernel() &&
+               (constant->scope == nullptr || parser::isa_node<parser::Cbase>(constant->scope.get()));
+    }
+
+    return false;
+}
+
+bool isRaise(parser::Node *node) {
+    auto raise = parser::cast_node<parser::Send>(node);
+
+    if (!raise) {
+        return false;
+    }
+
+    if (raise->method != core::Names::raise()) {
+        return false;
+    }
+
+    return raise->receiver == nullptr || isSelfOrKernel(raise->receiver.get());
+}
+
+void ensureAbstractMethodRaises(core::MutableContext ctx, const parser::Node *node) {
+    core::LocOffsets editLoc;
+    bool semicolon = false;
+    bool newline = false;
+
+    if (auto method = parser::cast_node<parser::DefMethod>((parser::Node *)node)) {
+        if (isRaise(method->body.get())) {
+            // If the method raises properly, we remove the body the body to not error later (see error 5019)
+            method->body = nullptr;
+            return;
+        }
+
+        auto lineStart = core::Loc::pos2Detail(ctx.file.data(ctx), method->declLoc.endPos()).line;
+        auto lineEnd = core::Loc::pos2Detail(ctx.file.data(ctx), method->loc.endPos()).line;
+
+        if (method->body) {
+            editLoc = method->body->loc;
+        } else if (lineStart == lineEnd) {
+            editLoc = method->declLoc.copyEndWithZeroLength();
+            semicolon = true;
+        } else {
+            editLoc = method->declLoc.copyEndWithZeroLength();
+            newline = true;
+        }
+    } else if (auto method = parser::cast_node<parser::DefS>((parser::Node *)node)) {
+        if (isRaise(method->body.get())) {
+            // If the method raises properly, we remove the body the body to not error later (see error 5019)
+            method->body = nullptr;
+            return;
+        }
+
+        auto lineStart = core::Loc::pos2Detail(ctx.file.data(ctx), method->declLoc.endPos()).line;
+        auto lineEnd = core::Loc::pos2Detail(ctx.file.data(ctx), method->loc.endPos()).line;
+
+        if (method->body) {
+            editLoc = method->body->loc;
+        } else if (lineStart == lineEnd) {
+            editLoc = method->declLoc.copyEndWithZeroLength();
+            semicolon = true;
+        } else {
+            editLoc = method->declLoc.copyEndWithZeroLength();
+            newline = true;
+        }
```

Comment:
The two methods types seem to be handled in exactly the same way. What do you think about pulling the definition out into a separate function that optionally returns the editLoc, semicolon, and newline values in the case of a failure?

Reply:
https://github.com/sorbet/sorbet/pull/8994